src/java/com/zimbra/cs/nginx/NginxLookupExtension.java

```diff
@@ -880,13 +880,23 @@ private DomainExternalRouteInfo getDomainExternalRouteInfo(ILdapContext zlc, Con
 
         private Server lookupUpstreamImapServer(NginxLookupRequest req) throws ServiceException {
             ImapLoadBalancingMechanism LBMech = ImapLoadBalancingMechanism.newInstance();
-            Server server = prov.getLocalServer();
-            String[] imapServerAddrs;
-            if (server != null) {
-                imapServerAddrs = server.getReverseProxyUpstreamImapServers();
-            } else {
-                imapServerAddrs = prov.getConfig().getReverseProxyUpstreamImapServers();
+            String[] imapServerAddrs = null;
+            Account acct = prov.get(AccountBy.name, req.user);
+            if (acct != null) {
+                imapServerAddrs = acct.getServer().getReverseProxyUpstreamImapServers();
+            }
+
+            if (imapServerAddrs == null || imapServerAddrs.length == 0){
```

Comment:
This logic is not correct. The correct logic is:

IF Account->getServer->reverseProxyUpstreamIMAPServers->notNullNorEmpty 
    THEN
         return Account->getServer->reverseProxyUpstreamIMAPServers
    ELSE
        return Account->getServer->getHostname
END IF